### ISSUE TITLE
discover() not using right interface - use netifaces?

### ISSUE THREAD

**Issue by ghcs27** [SoCo/SoCo#619]
Following up on #615: SoCo often doesn't find the right interface to use for UPnP discovery. As a way of working around this, we already have the `interface_addr` parameter, with which one can provide the interface address to use. However, this is not a satisfactory solution: For example, it means that the code from the readme code doesn't work for some people.

The official controller sends the discovery packets on all interfaces. Unfortunately, python provides no cross-platform way to get all available interfaces. The documentation suggests using netifaces:

> Note:
> There is no easy cross-platform way to find out the addresses of the
> local machine's network interfaces. You might try the
> netifaces module <https://pypi.python.org/pypi/netifaces> and some
> code like this:
>
>      >>> from netifaces import interfaces, AF_INET, ifaddresses
>      >>> data = [ifaddresses(i) for i in interfaces()]
>      >>> [d[AF_INET][0]['addr'] for d in data if d.get(AF_INET)]
>      ['127.0.0.1', '192.168.1.20']
>
> This should provide you with a list of values to try for
> interface_addr if you are having trouble finding your Sonos devices.

Netifaces works with OS X, Linux and Windows, so it is expected to cover the needs of most users. We could use it to discover the interface addresses ourselves, if the system is supported. So, the question is:

**Can we add netifaces as a dependency?**

Pro: It provides a valuable cross-platform API we cannot replace
Con: Having more dependencies is generally a bad thing
Con: It is still in major version zero, so it might have bugs or break backward compatibility
--> so are we, and getting the interfaces is a core feature which has been tested well.

**Comment by AndreMiras** [comment]
Also had this issue, bringing `docker0` interface down fixes it. See debug session below:
```
[andre:~/workspace/sonos] [venv] $ python -c 'import soco; soco.discover()'
Traceback (most recent call last):
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/urllib3/connection.py", line 141, in _new_conn
    (self.host, self.port), self.timeout, **extra_kw)
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/urllib3/util/connection.py", line 83, in create_connection
    raise err
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/urllib3/util/connection.py", line 73, in create_connection
    sock.connect(sa)
OSError: [Errno 113] No route to host

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/urllib3/connectionpool.py", line 601, in urlopen
    chunked=chunked)
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/urllib3/connectionpool.py", line 357, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/lib/python3.6/http/client.py", line 1239, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1285, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1234, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1026, in _send_output
    self.send(msg)
  File "/usr/lib/python3.6/http/client.py", line 964, in send
    self.connect()
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/urllib3/connection.py", line 166, in connect
    conn = self._new_conn()
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/urllib3/connection.py", line 150, in _new_conn
    self, "Failed to establish a new connection: %s" % e)
urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7fb5ce7815c0>: Failed to establish a new connection: [Errno 113] No route to host

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/requests/adapters.py", line 440, in send
    timeout=timeout
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/urllib3/connectionpool.py", line 639, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/urllib3/util/retry.py", line 388, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='172.17.0.89', port=1400): Max retries exceeded with url: /xml/ZoneGroupTopology1.xml (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fb5ce7815c0>: Failed to establish a new connectio
n: [Errno 113] No route to host',))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/soco/discovery.py", line 185, in discover
    return zone.visible_zones
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/soco/core.py", line 1019, in visible_zones
    self._parse_zone_group_state()
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/soco/core.py", line 932, in _parse_zone_group_state
    cache_timeout=5)['ZoneGroupState']
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/soco/services.py", line 787, in GetZoneGroupState
    return self.send_command('GetZoneGroupState', *args, **kwargs)
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/soco/services.py", line 459, in send_command
    args = self.compose_args(action, kwargs)
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/soco/services.py", line 334, in compose_args
    for action in self.actions:
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/soco/services.py", line 646, in actions
    self._actions = list(self.iter_actions())
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/soco/services.py", line 673, in iter_actions
    scpd_body = requests.get(self.base_url + self.scpd_url).content
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/requests/api.py", line 72, in get
    return request('get', url, params=params, **kwargs)
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/requests/api.py", line 58, in request
    return session.request(method=method, url=url, **kwargs)
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/requests/sessions.py", line 508, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/requests/sessions.py", line 618, in send
    r = adapter.send(request, **kwargs)
  File "/home/andre/workspace/sonos/venv/lib/python3.6/site-packages/requests/adapters.py", line 508, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='172.17.0.89', port=1400): Max retries exceeded with url: /xml/ZoneGroupTopology1.xml (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fb5ce7815c0>: Failed to establish a new connec
tion: [Errno 113] No route to host',))
```
Then if I `sudo ifconfig docker0 down` and try again it works:
```
python -c 'import soco; print(soco.discover())'
{SoCo("172.17.0.89")}
```

**Comment by berlincount** [comment]
I have the same issue, even when giving the network interface manually - the packets seem to come from the right network interface, though. They go to `239.255.255.250`, which might need multicast configuration?

**Comment by pwt** [comment]
I think this was fixed by #767. Please reopen if this is still an issue.
